### DISCUSSION TITLE
Same find query produce different results

### DISCUSSION THREAD

**Discussion by kennethsato** [mikro-orm/mikro-orm#2194] [Q&A] [answered]
**Describe the bug**
I'm trying to do a find query.
It seems like whatever find query I do for the first time after node.js was restarted works fine.
But after a couple finds it returns me incorrect results.
Some records do not match filtering criteria.

**Stack trace**
It get no stack trace, I'm able to get the resulting json, but it do not match filtering criterias.

**To Reproduce**```
I have a Comerciais repository.
I have "comercial" entity that may have multiple "enviaexibidor", which have one "emissora".
I'm building my filter using date values ("datainicio" and "datafim") and emissoras (string splitted by commas)

`
  public async findAll(
    datainicio: string,
    datafim: string,
    emissoras: string,
  ): Promise<Comercial[]> {
    let filter = {};
    if (datainicio && datafim) {
      filter = {
        enviaexibidor: {
          $and: [
            { allowed_at: { $gte: new Date(datainicio) } },
            { allowed_at: { $lte: new Date(datafim) } },
            { deleted_at: null },
          ],
          emissora: { siglapromo: { $in: emissoras.split(',') } },
        },
      };
    }

    const comerciais = await this.ormRepository.find(filter, {
      cache: false,
      fields: [
        'id',
        'titulo',
        'codmaterial',
        'categoria',
        'globoexpress',
        'statusdownload',
        'ptovenda',
        'duracao',
        'midia_id',
        'player_id',
        'created_at',
        'updated_at',
        'deleted_at',
        {
          player: ['nome'],
          enviaexibidor: [
            'id',
            'comercial_id',
            'emissora_id',
            'status',
            'statusconferencia',
            'usuarioconferencia_id',
            'allowed_at',
            'allowed_api_at',
            'created_at',
            'updated_at',
            'deleted_at',
            { emissora: ['id', 'nome', 'siglapromo'] },
            {
              usuarioconferencia: ['id', 'nome'],
            },
          ],
          midia: [
            'id',
            'avaliacaoeng',
            'obseng',
            'avaliador_id',
            'checked_at',
            'created_at',
            'filesize',
            { avaliador: ['id', 'nome'] },
          ],
        },
      ],
      orderBy: { enviaexibidor: { allowed_at: 'asc' } },
      populate: [
        'enviaexibidor',
        'enviaexibidor.emissora',
        'enviaexibidor.usuarioconferencia',
        'midia',
        'midia.avaliador',
        'player',
      ],
    });
    return comerciais;
  }`

The easiest way to reproduce the error was to:
Filter using a month as date's filter: from 01/08/2021 to 31/08/2021 and any emissora (eg: CTA, LDA, MGA). As mentioned, I get a correct result for the first find.
And then filter with only one day: from 19/08/2021 to 19/08/2021, keeping the same emissoras (CTA, LDA, MGA)
As a result, I get a total of 99 results, but ten of them do not match the date filtering criterias. I get result records whose dates are beyond 19/08/2021.

Then I restart the API and try the second query again.
And I get the correct result with 89 correct records.

```

**Expected behavior**
It should always return the correct result

**Additional context**
I disabled caching on configure, as it is not creating any JSONs at temp directory.
I converted my json results to attached XLSXs for easier comprehension.
Please check "R column" on the XLSX (autorizado_em, which is the allowed_at value)
On the incorrect result, I should not get the following rows: 4, 5, 75, 78, 82, 86, 88, 89, 98, 99

[Correct result after restart.xlsx](https://github.com/mikro-orm/mikro-orm/files/7154642/Correct.result.after.restart.xlsx)
[Incorrect result.xlsx](https://github.com/mikro-orm/mikro-orm/files/7154643/Incorrect.result.xlsx)

**Versions**

| Dependency | Version |
| - | - |
| node | 6.14.13|
| typescript | 4.2.4|
| mikro-orm | 5.0.0-dev.62 | also tried downgrading to 4.5.9 |

**Comment by B4nan** [comment, marked as answer]
You need to have dedicated contexts for each request.

- https://mikro-orm.io/docs/installation#request-context
- https://mikro-orm.io/docs/identity-map

And this example pretty much describes your problem:

https://mikro-orm.io/docs/identity-map#problem-2---unstable-response-of-api-endpoints

**Reply by kennethsato** [reply]
Hello Martin!
Thank you for the answer.
I tried to include RequestContext to our code, but I'm not sure where should I include it.

Thank you for the answer, Martin!

I think that's exactly the situation, but then we tried to add RequestContext to our code, but we couldn't figure where to add it.
I've seen you've mentioned that it may be hard to archive that using DI containers and that's our case.

We have our routes file, that call our Controller, that calls a Service that calls the Repository, that makes the query.
We tried to include it on the Repository, but we don't have next there. We also didn't want to send next all the way there.
And we could also put it on the Controller, but we do not have our Entity Manager there yet.
We are using tsyringe for the injections:

Would RequestContext be the best solution for us?


```
@injectable()
class RelatorioComerciaisRepository implements IRelatorioComerciaisRepository {
  private ormRepository: EntityRepository<Comercial>;
  private em: any;
  constructor(
    @inject('DatabaseProvider')
    private databaseProvider: IDatabaseProvider,
  ) {
    this.em = this.databaseProvider.getConnection().em;

    this.ormRepository = this.em.getRepository(Comercial);
  }
```

On the other hand, we tried also adding em.clear on our find.
And apparently it works, would that be a acceptable solution?


```
  public async findAll(
    datainicio: string,
    datafim: string,
    emissoras: string,
  ): Promise<Comercial[]> {
    this.em.clear();
    const comerciais = await this.ormRepository.find(filter, {
      cache: false,
    });
    return comerciais;
  }
```

Thank you!!

**Reply by B4nan** [reply]
It should be integreted via middleware ideally, I guess you still have something like express or fastify under your own abstractions, right?

Clearing like this won't work with concurrent requests and potentially will break a lot of things (e.g. lost updates as you could clear in the middle of some method), also you would not be able to flush in parallel. Clearing is for testing and for specific cases like error recovery where you know the state can be broken.

**Reply by kennethsato** [reply]
Hello Martin.
First of all, thank you for the answers!

As you suggested, we did create a middleware for RequestContext:
```
import { RequestContext } from '@mikro-orm/core';
import { NextFunction, Request, Response } from 'express';

export default async (
  request: Request,
  response: Response,
  next: NextFunction,
): Promise<any> => {
  RequestContext.create(
    [request.mysqlConnection.em, request.mongoConnection.em],
    next,
  );
};
```


We had a few problems with multiple DB connections, but this helped us out:
https://github.com/mikro-orm/mikro-orm/issues/872

But now we are getting "Too Many Connections" error, as following:
```
DriverException: Too many connections
    at MySqlExceptionConverter.convertException (C:\ADX\adxv2_backend\node_modul
es\@mikro-orm\core\platforms\ExceptionConverter.js:8:16)
    at MySqlExceptionConverter.convertException (C:\ADX\adxv2_backend\node_modul
es\@mikro-orm\mysql\node_modules\@mikro-orm\mysql-base\MySqlExceptionConverter.j
s:78:22)
    at MySqlDriver.convertException (C:\ADX\adxv2_backend\node_modules\@mikro-or
m\core\drivers\DatabaseDriver.js:202:54)
    at C:\ADX\adxv2_backend\node_modules\@mikro-orm\core\drivers\DatabaseDriver.
js:206:24
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at MySqlDriver.find (C:\ADX\adxv2_backend\node_modules\@mikro-orm\mysql\node
_modules\@mikro-orm\knex\AbstractSqlDriver.js:46:24)
    at SqlEntityManager.find (C:\ADX\adxv2_backend\node_modules\@mikro-orm\core\
EntityManager.js:93:25)

previous Error: Too many connections
    at Packet.asError (C:\ADX\adxv2_backend\node_modules\@mikro-orm\mysql\node_m
odules\mysql2\lib\packets\packet.js:712:17)
    at ClientHandshake.execute (C:\ADX\adxv2_backend\node_modules\@mikro-orm\mys
ql\node_modules\mysql2\lib\commands\command.js:28:26)
    at Connection.handlePacket (C:\ADX\adxv2_backend\node_modules\@mikro-orm\mys
ql\node_modules\mysql2\lib\connection.js:425:32)
    at PacketParser.onPacket (C:\ADX\adxv2_backend\node_modules\@mikro-orm\mysql
\node_modules\mysql2\lib\connection.js:75:12)
    at PacketParser.executeStart (C:\ADX\adxv2_backend\node_modules\@mikro-orm\m
ysql\node_modules\mysql2\lib\packet_parser.js:75:16)
    at Socket.<anonymous> (C:\ADX\adxv2_backend\node_modules\@mikro-orm\mysql\no
de_modules\mysql2\lib\connection.js:82:25)
    at Socket.emit (events.js:376:20)
    at Socket.emit (domain.js:532:15)
    at addChunk (internal/streams/readable.js:309:12)
    at readableAddChunk (internal/streams/readable.js:284:9)
    at Socket.Readable.push (internal/streams/readable.js:223:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:188:23)
    at TCP.callbackTrampoline (internal/async_hooks.js:134:14) {
  code: 'ER_CON_COUNT_ERROR',
  errno: 1040,
  sqlState: '',
  sqlMessage: 'Too many connections'
}
```

Could you enlighten us, please?

**Reply by B4nan** [reply]
i have no idea, if this is happening locally, maybe you do instantiate the orm on each request? where do you call `MikroORM.init()`? it should happen only on app start, forks do reuse the same connection (the `RequestContext.create()` call won't create new connections for sure)

**Reply by kennethsato** [reply]
We have figured our problem out.
In case anyone goes through the same:
Our database connect method was in our database constructor
So when middleware was trying to get the EntityManager, it was also creating a new connection:

```
export default async (
  request: Request,
  response: Response,
  next: NextFunction,
): Promise<any> => {
  RequestContext.create(
    [request.mysqlConnection.em, request.mongoConnection.em],
    next,
  );
};
```

We've just validated, if connection is already open do not open another one.

Thanks a lot for the answers @B4nan !!!
